### PR TITLE
Add configuration to RecordCollectors

### PR DESCRIPTION
**Pull request by rembiesa**
image

### CODE REVIEW

File: layouts/basic/modules/Settings/RecordCollector/Configuration.tpl [hunk 1]
```diff
@@ -0,0 +1,59 @@
+{*<!-- {[The file is published on the basis of YetiForce Public License 5.0 that can be found in the following directory: licenses/LicenseEN.txt or yetiforce.com]} -->*}
+<!-- tpl-Settings-RecordCollector-Configuration -->
+<div id="menuEditorContainer">
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Settings/RecordCollector/Configuration.tpl]: ?

File: layouts/basic/modules/Settings/RecordCollector/Configuration.tpl [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+						<thead>
+							<tr>
+								<th class="" scope="col">{\App\Language::translate('LBL_NAME', $MODULENAME)}</th>
+								<th class="" scope="col">{\App\Language::translate('LBL_DESCRIPTION', $MODULENAME)}</th>
+								<th class="text-center" scope="col">{\App\Language::translate('LBL_ACTIVE', $MODULENAME)}</th>
+								<th class="text-center" scope="col">{\App\Language::translate('LBL_ACTIONS', $MODULENAME)}</th>
+							</tr>
+						</thead>
+						<tbody>
+							{function UNSET_POPOVER}
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Settings/RecordCollector/Configuration.tpl]: ?

File: layouts/basic/modules/Settings/RecordCollector/Configuration.tpl [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+									</td>
+									<td class="text-center">
+										<input class="js-status-change" name="is_active" value="{$ITEM['name']}" type="checkbox" {if $ITEM['active']} checked {/if}>
+									</td>
+									<td class="text-center">
+										<button class="btn btn-outline-secondary btn-sm js-show-config-modal js-popover-tooltip mr-1" type="button" data-name="{$ITEM['instance']->label}"
+											data-content="{\App\Language::translate('LBL_CONFIG', $MODULENAME)}">
+											<span class="fas fa-cog"></span>
+										</button>
+										{* {if false}
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Settings/RecordCollector/Configuration.tpl]: ?

File: modules/Settings/RecordCollector/actions/ActiveAjax.php [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+ * RecordCollector active action class.
+ */
+class Settings_RecordCollector_ActiveAjax_Action extends Settings_Vtiger_Save_Action
+{
+	public function __construct()
+	{
+		parent::__construct();
+		$this->exposeMethod('changeStatus');
+	}
+
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/RecordCollector/actions/ActiveAjax.php]: ?doc

File: modules/Settings/RecordCollector/actions/ActiveAjax.php [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+				\vtlib\Link::addLink(0, 'EDIT_VIEW_RECORD_COLLECTOR', $collectorName, 'App\RecordCollectors\\' . $collectorName);
+			} elseif ('false' === $status) {
+				\vtlib\Link::deleteLink(0, 'EDIT_VIEW_RECORD_COLLECTOR', $collectorName);
+			}
+
+			$result = ['success' => true, 'message' => 'success_message'];
+			$response = new Vtiger_Response();
+			$response->setResult($result);
+			$response->emit();
+			return;
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/RecordCollector/actions/ActiveAjax.php]: ?

File: modules/Settings/RecordCollector/actions/ActiveAjax.php [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+{
+	public function __construct()
+	{
+		parent::__construct();
+		$this->exposeMethod('changeStatus');
+	}
+
+	public function changeStatus(App\Request $request): void
+	{
+		if ($request->get('collector') && $request->get('status')) {
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/RecordCollector/actions/ActiveAjax.php]: ?

File: modules/Settings/RecordCollector/actions/ActiveAjax.php [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+			$collectorName = $request->get('collector');
+			$status = $request->get('status');
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/RecordCollector/actions/ActiveAjax.php]:
```suggestion
			$status = $request->getBool('status');
```

File: modules/Settings/RecordCollector/actions/ActiveAjax.php [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+	public function __construct()
+	{
+		parent::__construct();
+		$this->exposeMethod('changeStatus');
+	}
+
+	public function changeStatus(App\Request $request): void
+	{
+		if ($request->get('collector') && $request->get('status')) {
+			$collectorName = $request->get('collector');
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/RecordCollector/actions/ActiveAjax.php]: getByType

File: modules/Settings/RecordCollector/actions/ActiveAjax.php [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+			}
+
+			$result = ['success' => true, 'message' => 'success_message'];
+			$response = new Vtiger_Response();
+			$response->setResult($result);
+			$response->emit();
+			return;
+		}
+
+		$result = ['success' => false, 'message' => 'error_message'];
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/RecordCollector/actions/ActiveAjax.php]: ?

File: modules/Settings/RecordCollector/actions/SaveConfig.php [hunk 10]
```diff
@@ -0,0 +1,27 @@
+<?php
+
+/**
+ * Settings RecordCollector SaveConfig action class.
+ *
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/RecordCollector/actions/SaveConfig.php:7]: @package

File: modules/Settings/RecordCollector/actions/SaveConfig.php [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+ * @copyright YetiForce S.A.
+ * @license YetiForce Public License 5.0 (licenses/LicenseEN.txt or yetiforce.com)
+ */
+class Settings_RecordCollector_SaveConfig_Action extends Settings_Vtiger_Basic_Action
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/RecordCollector/actions/SaveConfig.php:16]: class doc

File: modules/Settings/RecordCollector/models/Module.php [hunk 12]
```diff
@@ -0,0 +1,29 @@
+<?php
+
+/**
+ * Settings RecordCollector module model class.
+ *
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/RecordCollector/models/Module.php:7]: @package

File: modules/Settings/RecordCollector/models/Module.php [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+ * @copyright YetiForce S.A.
+ * @license YetiForce Public License 5.0 (licenses/LicenseEN.txt or yetiforce.com)
+ */
+class Settings_RecordCollector_Module_Model extends Settings_Vtiger_Module_Model
+{
+	private $collectors = [];
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/RecordCollector/models/Module.php:19]: doc

File: modules/Settings/RecordCollector/models/Module.php [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/RecordCollector/models/Module.php:20]: doc

File: modules/Settings/RecordCollector/models/Module.php [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+	public function getCollectors()
+	{
+		$active = (new \App\Db\Query())->select(['linklabel'])->from('vtiger_links')->where(['linktype' => 'EDIT_VIEW_RECORD_COLLECTOR'])->column();
+		$iterator = new \DirectoryIterator('App' . DIRECTORY_SEPARATOR . 'RecordCollectors');
+		foreach ($iterator as $item) {
+			if ($item->isFile() && 'php' === $item->getExtension() && 'Base' != $item->getBasename('.php')) {
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/RecordCollector/models/Module.php]: $file = $item->getBasename('.php')

File: modules/Settings/RecordCollector/views/Configuration.php [hunk 16]
```diff
@@ -0,0 +1,24 @@
+<?php
+
+/**
+ * Settings RecordCollector Configuration view class.
+ *
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/RecordCollector/views/Configuration.php:7]: @Package

File: modules/Settings/RecordCollector/views/Configuration.php [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+ * @copyright YetiForce S.A.
+ * @license YetiForce Public License 5.0 (licenses/LicenseEN.txt or yetiforce.com)
+ */
+class Settings_RecordCollector_Configuration_View extends Settings_Vtiger_Index_View
+{
+	public function preProcess(App\Request $request, $display = true)
+	{
+		parent::preProcess($request);
+	}
+
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/RecordCollector/views/Configuration.php]: doc

File: modules/Settings/RecordCollector/views/Configuration.php [hunk 18]
```diff
@@ -0,0 +1,24 @@
+<?php
+
+/**
+ * Settings RecordCollector Configuration view class.
+ *
+ * @copyright YetiForce S.A.
+ * @license YetiForce Public License 5.0 (licenses/LicenseEN.txt or yetiforce.com)
+ */
+class Settings_RecordCollector_Configuration_View extends Settings_Vtiger_Index_View
+{
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/RecordCollector/views/Configuration.php:17]: doc

File: modules/Settings/RecordCollector/views/Configuration.php [hunk 19]
```diff
@@ -0,0 +1,24 @@
...
+	public function preProcess(App\Request $request, $display = true)
+	{
+		parent::preProcess($request);
+	}
+
+	public function process(App\Request $request)
+	{
+		$viewer = $this->getViewer($request);
+
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/RecordCollector/views/Configuration.php]: ?

File: modules/Settings/RecordCollector/views/Configuration.php [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+		$viewer->assign('MODULENAME', $request->getModule(false));
+		$viewer->assign('COLLECTORS', Settings_RecordCollector_Module_Model::getInstance('Settings:RecordCollector')->getCollectors());
+		echo $viewer->view('Configuration.tpl', $request->getModule(false), true);
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/RecordCollector/views/Configuration.php]:
```suggestion
		$viewer->view('Configuration.tpl', $request->getModule(false));
```

**Review by mariuszkrzaczkowski — CHANGES_REQUESTED** [review]

File: layouts/basic/modules/Settings/RecordCollector/ConfigModal.tpl [hunk 21]
```diff
@@ -0,0 +1,41 @@
+{*<!-- {[The file is published on the basis of YetiForce Public License 5.0 that can be found in the following directory: licenses/LicenseEN.txt or yetiforce.com]} -->*}
+{strip}
+	<!-- tpl-Settings-RecrodCollector-RecordCollectorConfigModal -->
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Settings/RecordCollector/ConfigModal.tpl]:
```suggestion
	<!-- tpl-Settings-RecrodCollector-ConfigModal-->
```

File: layouts/basic/modules/Settings/RecordCollector/ConfigModal.tpl [hunk 22]
```diff
@@ -0,0 +1,41 @@
...
+									</td>
+								</tr>
+							{/foreach}
+						</tbody>
+					</table>
+				</div>
+			</div>
+		</form>
+	</div>
+	<!-- /tpl-Settings-RecrodCollector-RecordCollectorConfigModal -->
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Settings/RecordCollector/ConfigModal.tpl]:
```suggestion
	<!-- /tpl-Settings-RecrodCollector-ConfigModal-->
```

**Comment by rembiesa** [comment]
image

File: modules/Settings/RecordCollector/actions/SaveAjax.php [hunk 23]
```diff
@@ -0,0 +1,46 @@
+<?php
+/**
+ * RecordCollector  active action file.
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/RecordCollector/actions/SaveAjax.php]: 2 space

File: modules/Settings/RecordCollector/actions/SaveAjax.php [hunk 24]
```diff
@@ -0,0 +1,46 @@
...
+ * @license   YetiForce Public License 5.0 (licenses/LicenseEN.txt or yetiforce.com)
+ * @author    Sławomir Rembiesa <s.rembiesa@yetiforce.com>
+ */
+
+/**
+ * RecordCollector active action class.
+ */
+class Settings_RecordCollector_SaveAjax_Action extends Settings_Vtiger_Save_Action
+{
+	public function __construct()
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/RecordCollector/actions/SaveAjax.php:20]: doc

File: modules/Settings/RecordCollector/actions/SaveAjax.php [hunk 25]
```diff
@@ -0,0 +1,46 @@
...
+	 * @param App\Request $request
+	 *
+	 * @return void
+	 */
+	public function changeStatus(App\Request $request): void
+	{
+		$collectorName = $request->getByType('collector');
+		$status = $request->getBoolean('status');
+
+		if ($status) {
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/RecordCollector/actions/SaveAjax.php]:
```suggestion
		if ($request->getBoolean('status')) {
```

File: modules/Settings/RecordCollector/actions/SaveAjax.php [hunk 26]
```diff
@@ -0,0 +1,46 @@
...
+			\vtlib\Link::addLink(0, 'EDIT_VIEW_RECORD_COLLECTOR', $collectorName, 'App\RecordCollectors\\' . $collectorName);
+		} elseif (false === $status) {
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/RecordCollector/actions/SaveAjax.php]:
```suggestion
		} else {
```

File: modules/Settings/RecordCollector/actions/SaveAjax.php [hunk 27]
```diff
@@ -0,0 +1,46 @@
...
+			\vtlib\Link::deleteLink(0, 'EDIT_VIEW_RECORD_COLLECTOR', $collectorName);
+		}
+
+		$result = ['success' => true, 'message' => 'success_message'];
+		$response = new Vtiger_Response();
+		$response->setResult($result);
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/RecordCollector/actions/SaveAjax.php]:
```suggestion
		$response->setResult(['success' => true, 'message' => 'success_message']);
```

File: modules/Settings/RecordCollector/views/CollectorConfigModal.php [hunk 28]
```diff
@@ -0,0 +1,36 @@
+<?php
+
+/**
+ * YetiForce product Modal.
+ *
+ * @package Settings.View
+ *
+ * @copyright YetiForce S.A.
+ * @license   YetiForce Public License 5.0 (licenses/LicenseEN.txt or yetiforce.com)
+ * @author    Tomasz Poradzewski <t.poradzewski@yetiforce.com>
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/RecordCollector/views/CollectorConfigModal.php]: ?

File: modules/Settings/RecordCollector/views/CollectorConfigModal.php [hunk 29]
```diff
@@ -0,0 +1,36 @@
...
+ */
+
+/**
+ * Offline registration modal view class.
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/RecordCollector/views/CollectorConfigModal.php]: ?

File: modules/Settings/RecordCollector/views/CollectorConfigModal.php [hunk 30]
```diff
@@ -0,0 +1,36 @@
+<?php
+
+/**
+ * YetiForce product Modal.
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/RecordCollector/views/CollectorConfigModal.php]: ?

File: modules/Settings/RecordCollector/views/CollectorConfigModal.php [hunk 31]
```diff
@@ -0,0 +1,36 @@
...
+
+	/** {@inheritdoc} */
+	public function process(App\Request $request)
+	{
+		$viewer = $this->getViewer($request);
+		$provider = \App\Map\Address::getInstance($request->getByType('provider'));
+		$viewer->assign('CONFIG', Settings_ApiAddress_Module_Model::getInstance('Settings:ApiAddress')->getConfig()[$provider->getName()] ?? []);
+		$viewer->assign('PROVIDER', $provider);
+		$viewer->assign('CUSTOM_FIELDS', $provider->getCustomFields());
+		$viewer->view('ApiConfigModal.tpl', $this->qualifiedModuleName);
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/RecordCollector/views/CollectorConfigModal.php]:
```suggestion
		$viewer->view('ConfigModal.tpl', $this->qualifiedModuleName);
```

File: modules/Settings/RecordCollector/views/Configuration.php [hunk 32]
```diff
@@ -0,0 +1,32 @@
...
+		parent::preProcess($request);
+	}
+
+	/** {@inheritdoc} */
+	public function process(App\Request $request)
+	{
+		$viewer = $this->getViewer($request);
+		$viewer->assign('MODULENAME', $request->getModule(false));
+		$viewer->assign('COLLECTORS', Settings_RecordCollector_Module_Model::getInstance('Settings:RecordCollector')->getCollectors());
+		echo $viewer->view('List.tpl', $request->getModule(false));
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/RecordCollector/views/Configuration.php]:
```suggestion
		$viewer->view('List.tpl', $request->getModule(false));
```

**Review by mariuszkrzaczkowski — CHANGES_REQUESTED** [review]

**Comment by rembiesa** [comment]
image

File: modules/Settings/RecordCollector/models/Module.php [hunk 33]
```diff
@@ -0,0 +1,43 @@
...
+{
+	/** @var array collectors */
+	private $collectors = [];
+
+	/**
+	 * Function fetching all collectors in system.
+	 *
+	 * @return array
+	 */
+	public function getCollectors()
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/RecordCollector/models/Module.php]:
```suggestion
	public function getCollectors(): array
```

File: modules/Settings/RecordCollector/views/Configuration.php [hunk 34]
```diff
@@ -0,0 +1,32 @@
...
+ * @author Sławomir Rembiesa <s.rembiesa@yetiforce.com>
+ */
+
+/**
+ * Settings RecordCollector Configuration view class.
+ */
+class Settings_RecordCollector_Configuration_View extends Settings_Vtiger_Index_View
+{
+	/** {@inheritdoc} */
+	public function preProcess(App\Request $request, $display = true)
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/RecordCollector/views/Configuration.php]: ?

**Review by mariuszkrzaczkowski — CHANGES_REQUESTED** [review]

**Pull request merged**